// @ngx-loading-bar/http/esm2015/ngx-loading-bar-http.js
import { LoadingBarService, LoadingBarModule } from '@ngx-loading-bar/core';
import { Injectable, NgModule } from '@angular/core';
import { ConnectionBackend, Http, RequestOptions, HttpModule, XHRBackend } from '@angular/http';
import { finalize } from 'rxjs/operators';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */

class LoadingBarHttp extends Http {
    /**
     * @param {?} _backend
     * @param {?} _defaultOptions
     * @param {?} loadingBar
     */
    constructor(_backend, _defaultOptions, loadingBar) {
        super(_backend, _defaultOptions);
        this.loadingBar = loadingBar;
    }
    /**
     * @param {?} url
     * @param {?=} options
     * @return {?}
     */
    request(url, options) {
        let /** @type {?} */ headers = options && options.headers;
        if (!headers && url && (/** @type {?} */ (url)).headers) {
            headers = (/** @type {?} */ (url)).headers;
        }
        if (headers && headers.has('ignoreLoadingBar')) {
            headers.delete('ignoreLoadingBar');
            return super.request(url, options);
        }
        const /** @type {?} */ response$ = super.request(url, options);
        if (options && options.ignoreLoadingBar === true) {
            return response$;
        }
        let /** @type {?} */ started = false;
        const /** @type {?} */ responseSubscribe = response$.subscribe.bind(response$);
        response$.subscribe = (...args) => {
            this.loadingBar.start();
            started = true;
            return responseSubscribe(...args);
        };
        return response$.pipe(finalize(() => started && this.loadingBar.complete()));
    }
}
LoadingBarHttp.decorators = [
    { type: Injectable },
];
/** @nocollapse */
LoadingBarHttp.ctorParameters = () => [
    { type: ConnectionBackend, },
    { type: RequestOptions, },
    { type: LoadingBarService, },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @param {?} xhrBackend
 * @param {?} requestOptions
 * @param {?} loadingBarService
 * @return {?}
 */
function httpFactory(xhrBackend, requestOptions, loadingBarService) {
    return new LoadingBarHttp(xhrBackend, requestOptions, loadingBarService);
}
class LoadingBarHttpModule {
}
LoadingBarHttpModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpModule,
                    LoadingBarModule.forRoot(),
                ],
                exports: [
                    HttpModule,
                    LoadingBarModule,
                ],
                providers: [
                    { provide: Http, useFactory: httpFactory, deps: [XHRBackend, RequestOptions, LoadingBarService] },
                ],
            },] },
];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */

export { LoadingBarHttpModule, LoadingBarHttp, httpFactory as ɵa };
